File: Bag3/Code/Stations/Icon Display Station/iconlib/preview.py
"""
preview.py -- render a 16x16 authored ICON tuple as a simulated-LED PNG,
through the SAME forward model (§Root cause) the editor's live canvas
uses, so it stops lying the way the raw-value ANSI preview used to.

Honest about what it's good for: cell-on/off, silhouette, thin-feature
survival, relative brightness ordering, quantization damage. NOT reliable
for absolute hue -- WS2812 dies are narrowband/off-primary in a way no
sRGB monitor simulation captures. Sign off on structure here; sign off on
hue only from the real matrix.
"""

from PIL import Image, ImageDraw, ImageFilter

from ledcolor import predict_led_appearance
from raster import W, H

PREVIEW_SCALE = 24
DOT_RADIUS = 8
BLOOM_RADIUS = 3
BG = (10, 10, 12)


def render_preview(pixels, intensity, out_path=None):
    """pixels: flat 256 (r,g,b) authored tuples, row-major top-left.
    Returns a PIL Image; also writes it to out_path if given."""
    size = W * PREVIEW_SCALE
    img = Image.new("RGB", (size, size), BG)
    dots = Image.new("RGB", (size, size), (0, 0, 0))
    draw = ImageDraw.Draw(dots)

    for row in range(H):
        for col in range(W):
            rgb = pixels[row * W + col]
            if rgb == (0, 0, 0):
                continue
            seen = predict_led_appearance(rgb, intensity)
            cx = col * PREVIEW_SCALE + PREVIEW_SCALE // 2
            cy = row * PREVIEW_SCALE + PREVIEW_SCALE // 2
            draw.ellipse((cx - DOT_RADIUS, cy - DOT_RADIUS, cx + DOT_RADIUS, cy + DOT_RADIUS), fill=seen)

    bloom = dots.filter(ImageFilter.GaussianBlur(BLOOM_RADIUS))
    out = Image.blend(img, bloom, 0.35)
    # simple screen-blend of the sharp dots over the bloomed background, clamped
    px_out = out.load()
    px_dots = dots.load()
    for y in range(size):
        for x in range(size):
            bo = px_out[x, y]
            do = px_dots[x, y]
            px_out[x, y] = tuple(min(255, bo[c] + do[c]) for c in range(3))

    if out_path:
        out.save(out_path)
    return out

File: Live_Page/ChatBroadcast/iconmaker/js/pipeline/preview.js
/**
 * preview.js -- simulated-LED render, ported from iconlib/preview.py but
 * using three composited canvases instead of Pillow's per-pixel additive
 * loop (see plan §Preview compose for why plain shadowBlur under
 * source-over is wrong: it occludes overlapping glows instead of summing
 * them, the way Python's whole-layer Gaussian + clamped add does).
 *
 * Honest about what it's good for: cell on/off, silhouette, thin-feature
 * survival, relative brightness ordering -- NOT absolute hue. Sign off on
 * hue only from the real matrix (see readme.md).
 */

import { W, H, PREVIEW_SCALE, DOT_RADIUS, BLOOM_RADIUS, PREVIEW_BG } from "./constants.js";
import { predictLedAppearance } from "./ledcolor.js";

const isBlack = (p) => p[0] === 0 && p[1] === 0 && p[2] === 0;

/**
 * @param {Array<[number,number,number]>} pixels  flat 256, row-major, authored bytes
 * @param {number} intensity
 * @param {HTMLCanvasElement} [canvas]  reused if given, else created
 * @returns {HTMLCanvasElement}
 */
export function renderPreview(pixels, intensity, canvas) {
  const size = W * PREVIEW_SCALE;
  canvas = canvas || document.createElement("canvas");
  canvas.width = size;
  canvas.height = size;
  const ctx = canvas.getContext("2d");

  const dots = document.createElement("canvas");
  dots.width = size;
  dots.height = size;
  const dctx = dots.getContext("2d");
  dctx.fillStyle = "rgb(0, 0, 0)";
  dctx.fillRect(0, 0, size, size);
  dctx.globalCompositeOperation = "lighter"; // sums overlapping glows, matches Python's additive clamp

  for (let row = 0; row < H; row++) {
    for (let col = 0; col < W; col++) {
      const rgb = pixels[row * W + col];
      if (isBlack(rgb)) continue;
      const seen = predictLedAppearance(rgb, intensity);
      const cx = col * PREVIEW_SCALE + PREVIEW_SCALE / 2;
      const cy = row * PREVIEW_SCALE + PREVIEW_SCALE / 2;
      dctx.fillStyle = `rgb(${seen[0]}, ${seen[1]}, ${seen[2]})`;
      dctx.beginPath();
      dctx.arc(cx, cy, DOT_RADIUS, 0, Math.PI * 2);
      dctx.fill();
    }
  }

  const bloom = document.createElement("canvas");
  bloom.width = size;
  bloom.height = size;
  const bctx = bloom.getContext("2d");
  bctx.filter = `blur(${BLOOM_RADIUS}px)`; // true Gaussian, stdDev == BLOOM_RADIUS px
  bctx.drawImage(dots, 0, 0);

  ctx.globalCompositeOperation = "source-over";
  ctx.fillStyle = `rgb(${PREVIEW_BG[0]}, ${PREVIEW_BG[1]}, ${PREVIEW_BG[2]})`;
  ctx.fillRect(0, 0, size, size);
  ctx.globalAlpha = 0.35; // Image.blend(bg, bloom, 0.35)
  ctx.drawImage(bloom, 0, 0);
  ctx.globalAlpha = 1.0;
  ctx.globalCompositeOperation = "lighter"; // clamped additive sharp layer on top
  ctx.drawImage(dots, 0, 0);
  ctx.globalCompositeOperation = "source-over";

  return canvas;
}

export function previewToBlob(canvas) {
  return new Promise((resolve) => canvas.toBlob(resolve, "image/png"));
}
